andela/boondocks-bn-backend
# #169257398  Login in via Facebook and Google

#### What does this PR do?
Implements the user login using Facebook or Google user information
#### Description of Task to be completed?
- [x] Register our Application to get `Facebook App Id` and `Facebook Client Secret`
- [x] Register our Application to get the `Google Client Id` and `Google Client Secret`
These details help facebook and google respectively to know which application is requesting access, as explained on facebook developer [website](https://developers.facebook.com/docs/apps/)
> Registration creates an App ID that lets us know who you are, helps us distinguish your app from other apps, and provides a way for you to supply any additional materials we may need when you set up specific products or request access to sensitive APIs. 
- [x] Configured the whitelisted URL for both the Facebook and Google Account to protect our app from being blocked with CORS (Which will be modified when in production as the url adress will be changed)
- Google: 
`http://localhost:3000/api/v1/auth/google`, 
`http://localhost:3030/api/v1/auth/google`
- Facebook: 
`http://localhost:3000/api/v1/auth/facebook`, 
`http://localhost:3030/api/v1/auth/facebook`
- Created a `/api/v1/auth/access` route (this will be changed to the dashboard route later) to which the user will be redirected with user object response, when he authenticates well

#### How should this be manually tested?
- Clone the repo to the local machine
- Install the dependencies by running `npm install` from the app root folder
- Run `npm run start` to start the application
- Run `npm run migrate` to create the database table where the user object will be created
- In the browser navigate to `http://localhost:3000/api/v1/auth/google` if you want to grant the application access to your Google user info or to `http://localhost:3000/api/v1/auth/facebook` if you need it to access to your Facebook user info. Then you'll be redirected to the google or facebook websites respectively, and be asked to allow the application to access one of your emails' information
- Click on the email you want the application to have access to (For Google), or on the blue button labeled `Continue as ...`, and you will be redirected back to our application, and in your browser, you will see a JSON object containing the `status`, `message` and `token`, and this token will be used to log the user in our application.
- Upon completion of the information retrieval from Google or Facebook, when the user with the same email as the one present in the object returned was found, that user will be logged in directly otherwise a new user will be created and he will be logged in as well.
#### Any background context you want to provide?
N/A
#### What are the relevant pivotal tracker stories?
[#169257398](https://www.pivotaltracker.com/story/show/169257398)
#### Screenshots (if appropriate)r
_Google screen to grant access on user info to our application:_
![image](https://user-images.githubusercontent.com/3137192/68991633-4cb8f700-0869-11ea-94d0-18fbaa94bd66.png)

_Facebook screen to grant access on user info to our application:_
![image](https://user-images.githubusercontent.com/3137192/68992202-1ed7b080-0871-11ea-8f16-e8bd4867b481.png)


_Object returned when tested in the browser_

```
{
  "status": "success",
  "message": "User logged in successfully",
  "data": {
    "id": 1,
    "firstName": "Gildas",
    "lastName": "Niyigena",
    "email": "gildniy05@gmail.com",
    "token": "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJlbWFpbCI6ImdpbGRuaXkwNUBnbWFpbC5jb20iLCJ1c2VySWQiOjEsInZlcmlmaWVkIjp0cnVlLCJpYXQiOjE1NzM5MDE0OTYsImV4cCI6MTU3Mzk4Nzg5Nn0.NAHGN8U1dytC-mjGFDMWEoIyaqXj5ZoUQeuEym1dvAQ"
  }
}
```

#### Questions:

1. As for this feature, the user will be redirected to another website (Google or Facebook) for granting our application the access to his info and returns back to our application with the needed data; this cannot be tested in Postman (because the redirections that happen). Can the browser be accepted as a way to test this feature and not Postman application?

2. With the new user to the application, created by this method, we already know that he owns the email and other info as these are being provided by Google and Facebook, we are sure they have verified he is. I wonder if the fact that I set him as verified directly when logging in with these platforms is a bad idea.

3. Is there a way to test the response generated as they are from a redirection (from Facebook or Gmail); and the code provided in the URL to be handle can't be mocked (It's a 1 time usable and is only generated by Facebook and Google).
